### PR TITLE
Add V2 Route for Web Actions

### PR DESCRIPTION
**Pull request by dubeejw**
- Add V2 route that supports only Web Actions while leaving V1 route for Web Actions intact

### CODE REVIEW

**Comment by dubeejw** [comment]
WIP

**Comment by dubeejw** [comment]
@rabbah, @csantanapr, just to confirm, we want Web Actions to be accessible from the v2 route instead of v1?

**Comment by rabbah** [comment]
@dubeejw consult this issue: https://github.com/openwhisk/openwhisk/issues/1965.
I would leave api/v1 alone, and start api/v2 of which the only route available is web actions.

File: core/controller/src/main/scala/whisk/core/controller/Meta.scala [hunk 1]
```diff
@@ -217,13 +217,10 @@ trait WhiskMetaApi
     /** Store for identities. */
     protected val authStore: AuthStore
 
-    /** The route prefix e.g., /experimental. */
-    protected val routePath = "experimental"
```
Comment (dubeejw) [inline core/controller/src/main/scala/whisk/core/controller/Meta.scala]: This is a problem if we want to support `v1/experimental/web` and `v2/web` simultaneously.

Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Meta.scala]: Why? It's the same code, just two different routes? Let's discuss on slack.

File: core/controller/src/main/scala/whisk/core/controller/Controller.scala [hunk 2]
```diff
@@ -67,14 +67,15 @@ class Controller(
...
-            super.routes ~ apiv1.routes
+            super.routes ~ apiv1.routes ~ apiv2.routes
         }
     }
 
     logging.info(this, s"starting controller instance ${instance}")
 
     /** The REST APIs. */
     private val apiv1 = new RestAPIVersion_v1(config, context.system, logging)
+    private val apiv2 = new RestAPIVersion_v2(config, context.system, logging)
```
Comment (dubeejw) [inline core/controller/src/main/scala/whisk/core/controller/Controller.scala]: Currently seeing a pause when calling which ever route is instantiated second here...

Comment (dubeejw) [inline core/controller/src/main/scala/whisk/core/controller/Controller.scala]: @rabbah, any idea why the above behavior would occur?

File: tests/src/test/scala/whisk/core/controller/test/MetaApiTests.scala [hunk 3]
```diff
@@ -64,16 +65,13 @@ import whisk.http.Messages
  * "using Specification DSL to write unit tests, as in should, must, not, be"
  * "using Specs2RouteTest DSL to chain HTTP requests for unit testing, as in ~>"
  */
-@RunWith(classOf[JUnitRunner])
-class MetaApiTests extends ControllerTestCommon with WhiskMetaApi with BeforeAndAfterEach {
-
-    override val apipath = "api"
-    override val apiversion = "v1"
```
Comment (dubeejw) [inline tests/src/test/scala/whisk/core/controller/test/MetaApiTests.scala]: Having trouble overriding these variables at the proper time during instantiation, specifically for webInvokePath. Hence the lazy val technique.

**Comment by dubeejw** [comment]
Just for fun, PG2 1167.

**Comment by dubeejw** [comment]
@rabbah, please review.

**Comment by dubeejw** [comment]
PG2 1188, approved.

File: core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala [hunk 4]
```diff
@@ -94,6 +95,21 @@ protected[controller] object RestAPIVersion_v1 {
 }
 
 /**
+  * A singleton object which defines properties needed to instantiate a service for v2
+  * of the REST API.
+  */
+protected[controller] object RestAPIVersion_v2 {
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: this is the same as v1 right?
how about sharing them?

File: core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala [hunk 5]
```diff
@@ -274,6 +282,9 @@ protected[controller] class RestAPIVersion_v1(
             override val logging: Logging)
         extends WhiskMetaApi with WhiskServices {
         override val whiskConfig = config
+        override lazy val apipath = "api"
+        override lazy val apiversion = "v1"
+        override lazy val webInvokePath = "experimental" / "web"
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: why is this hard coded here but passed in as a parameter above?

File: core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala (hunk 4 above)
Comment (dubeejw) [inline core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: I was able to share the MetasApi class. RestAPIVersion_v1 and RestAPIVersion_v2 are a little different unless you have suggestions.

File: core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala [hunk 6]
```diff
@@ -290,3 +266,63 @@ protected[controller] class RestAPIVersion_v1(
...
+
+    /**
+      * Here is the key method: it defines the Route (route tree) which implement v2 of the REST API.
+      *
+      * @Idioglossia This relies on the spray routing DSL.
+      * @see http://spray.io/documentation/1.2.2/spray-routing/
+      */
+    override def routes(implicit transid: TransactionId): Route = {
+        pathPrefix(apipath / apiversion) {
+            sendCorsHeaders {
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/RestAPIs.scala]: nuke `sendCorsHeaders { `

**Comment by rabbah** [comment]
superseded by #2004.

**Pull request closed without merging**